### PR TITLE
Make sure onPublishSuccess is called during cluster state tests

### PR DESCRIPTION
**Pull request by thecoop**
Make sure `onPublishSuccess` is actually called in test scenarios. This is to help with testing of #101609

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/cluster/service/ClusterStateTaskExecutorUtils.java [hunk 1]
```diff
@@ -123,6 +123,7 @@ public void success(Runnable onPublishSuccess, ClusterStateAckListener clusterSt
             assert clusterStateAckListener != null;
             assert task == clusterStateAckListener || (task instanceof ClusterStateAckListener == false);
             this.succeeded = true;
+            onPublishSuccess.run();
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/cluster/service/ClusterStateTaskExecutorUtils.java:126]: Ah no we shouldn't run `onPublishSuccess` here, we're still executing the update at this point. We should hold on to the success callback and run it once the update is fully computed and "published" (whatever that might mean in these tests).

File: server/src/test/java/org/elasticsearch/ingest/IngestServiceTests.java [hunk 2]
```diff
@@ -2680,7 +2680,19 @@ private static void executeFailingDelete(DeletePipelineRequest request, ClusterS
...
+
+    private static <T> ActionListener<T> createTestListener() {
+        return new ActionListener<>() {
+            @Override
+            public void onResponse(T t) {}
+
+            @Override
+            public void onFailure(Exception e) {
+                throw new AssertionError(e);
+            }
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/ingest/IngestServiceTests.java:2694]: Suggest `org.elasticsearch.action.support.ActionTestUtils#assertNoFailureListener` for all these listeners that assert they don't fail.

File: test/framework/src/main/java/org/elasticsearch/cluster/service/ClusterStateTaskExecutorUtils.java (hunk 1 above)
Comment (thecoop) [reply test/framework/src/main/java/org/elasticsearch/cluster/service/ClusterStateTaskExecutorUtils.java:126]: But there's nothing actually to do here, so the update is trivially complete anyway?

Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/cluster/service/ClusterStateTaskExecutorUtils.java:126]: Hm that shouldn't be the case always, we might be simulating running a batch of tasks, the first of which succeeds (and gets here) and then the second throws an exception, aborting the whole batch.

Comment (thecoop) [reply test/framework/src/main/java/org/elasticsearch/cluster/service/ClusterStateTaskExecutorUtils.java]: Ah right, I've added a capture-and-call mechanism instead

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**